File: java-sdk/sdk/src/test/kotlin/org/apache/airflow/sdk/ConfigTest.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.airflow.sdk

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Files
import java.nio.file.Path

class ConfigTest {
  // -- SdkConfig: env var resolution --

  @Test
  @DisplayName("executionApiUrl uses AIRFLOW__CORE__EXECUTION_API_SERVER_URL env var")
  fun executionApiUrlFromEnv() {
    val config = SdkConfig(env = mapOf("AIRFLOW__CORE__EXECUTION_API_SERVER_URL" to "http://127.0.0.1:8080/execution"))

    assertEquals("http://127.0.0.1:8080/execution/", config.executionApiUrl)
  }

  @Test
  @DisplayName("executionApiUrl throws when missing")
  fun executionApiUrlThrowsWhenMissing() {
    val config = SdkConfig(env = emptyMap())

    val error = assertThrows(WorkerError::class.java) { config.executionApiUrl }
    assertTrue(error.message!!.contains("execution_api_server_url"))
  }

  @Test
  @DisplayName("executionApiUrl falls back to execution.api_url")
  fun executionApiUrlFallback() {
    val config = SdkConfig(env = mapOf("AIRFLOW__EXECUTION__API_URL" to "http://127.0.0.1:8080/execution"))

    assertEquals("http://127.0.0.1:8080/execution/", config.executionApiUrl)
  }

  @Test
  @DisplayName("jwtExpirationTime defaults to 30 seconds")
  fun jwtExpirationTimeDefault() {
    val config = SdkConfig(env = emptyMap())

    assertEquals(30, config.jwtExpirationTime)
  }

  // -- SdkConfig: YAML resolution --

  @Test
  @DisplayName("config values are loaded from YAML file")
  fun yamlConfigLoading(
    @TempDir tempDir: Path,
  ) {
    val yamlContent =
      """
      core:
        execution_api_server_url: "http://yaml-host:8080/execution/"

      sdk:
        bundles_dir: "./bundles"

      api_auth:
        jwt_secret: "yaml-secret"
        jwt_issuer: "yaml-issuer"
        jwt_expiration_time: 45
      """.trimIndent()

    val yamlPath = tempDir.resolve("java-sdk.yaml")
    Files.writeString(yamlPath, yamlContent)

    val config = SdkConfig(env = emptyMap(), yamlOverride = yamlPath)

    assertEquals("http://yaml-host:8080/execution/", config.executionApiUrl)
    assertEquals("yaml-secret", config.jwtSecret)
    assertEquals("yaml-issuer", config.jwtIssuer)
    assertEquals(45, config.jwtExpirationTime)
    assertEquals(Path.of("./bundles"), config.bundlesDir)
  }

  @Test
  @DisplayName("env vars take precedence over YAML values")
  fun envTakesPrecedenceOverYaml(
    @TempDir tempDir: Path,
  ) {
    val yamlContent =
      """
      core:
        execution_api_server_url: "http://yaml-host:8080/execution/"
      api_auth:
        jwt_secret: "yaml-secret"
      """.trimIndent()

    val yamlPath = tempDir.resolve("java-sdk.yaml")
    Files.writeString(yamlPath, yamlContent)

    val config =
      SdkConfig(
        env =
          mapOf(
            "AIRFLOW__CORE__EXECUTION_API_SERVER_URL" to "http://env-host:9090/execution/",
            "AIRFLOW__API_AUTH__JWT_SECRET" to "env-secret",
          ),
        yamlOverride = yamlPath,
      )

    assertEquals("http://env-host:9090/execution/", config.executionApiUrl)
    assertEquals("env-secret", config.jwtSecret)
  }

  @Test
  @DisplayName("config works with no YAML file and no env vars for optional values")
  fun noYamlFile() {
    val config = SdkConfig(env = emptyMap())

    assertEquals(30, config.jwtExpirationTime)
    assertNull(config.bundlesDir)
  }

  @Test
  @DisplayName("YAML file is resolved from AIRFLOW_HOME")
  fun yamlFromAirflowHome(
    @TempDir tempDir: Path,
  ) {
    val yamlContent =
      """
      api_auth:
        jwt_secret: "home-secret"
      """.trimIndent()

    Files.writeString(tempDir.resolve("java-sdk.yaml"), yamlContent)

    val config = SdkConfig(env = mapOf("AIRFLOW_HOME" to tempDir.toString()))

    assertEquals("home-secret", config.jwtSecret)
  }
}

File: airflow-core/src/airflow/ui/src/i18n/config.test.ts
/*!
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { createInstance } from "i18next";
import { describe, expect, it } from "vitest";

import { convertDetectedLanguage, i18nBaseOptions } from "./config";

// getBestMatchFromCodes is the resolver i18next runs on the array the
// LanguageDetector returns. It is not part of i18next's public types
// (services.languageUtils is `any`), so we pin the one method we use.
type LanguageUtils = { getBestMatchFromCodes: (codes: ReadonlyArray<string>) => string };

// Resolve a language exactly as the running app does: the detector maps
// convertDetectedLanguage over navigator.languages (preserving order), then
// i18next picks the best match. Uses the production init options so the test
// guards the real config rather than a copy.
const resolveLanguage = async (navigatorLanguages: ReadonlyArray<string>): Promise<string> => {
  const instance = createInstance();

  await instance.init({ ...i18nBaseOptions, initImmediate: false, resources: {} });

  const languageUtils = instance.services.languageUtils as LanguageUtils;

  return languageUtils.getBestMatchFromCodes(navigatorLanguages.map(convertDetectedLanguage));
};

describe("i18n language resolution", () => {
  it.each([
    // The reported bug: real UK Chrome with Hindi added. Before the fix this
    // resolved to "hi" because "en-GB"/"en-US" are not exact members of
    // supportedLngs but "hi" is, so the first bare supported code won.
    { expected: "en", languages: ["en-GB", "hi-IN", "hi", "en-US", "en", "gu"] },
    { expected: "en", languages: ["en-GB", "hi"] },
    { expected: "en", languages: ["en-GB"] },
    // Browsers may emit non-canonical casing.
    { expected: "en", languages: ["EN-GB", "hi"] },
    // Other region-qualified base languages were affected the same way.
    { expected: "pt", languages: ["pt-BR", "en"] },
    { expected: "de", languages: ["de-AT", "de", "en"] },
    { expected: "fr", languages: ["fr-FR", "fr", "en"] },
  ])(
    "resolves $languages to $expected (region variant of a base language)",
    async ({ expected, languages }) => {
      expect(await resolveLanguage(languages)).toBe(expected);
    },
  );

  it.each([
    // Region-specific supported locales must stay intact even when English
    // follows them -- this is what nonExplicitSupportedLngs/load:"languageOnly"
    // would break by collapsing "zh-CN"/"zh-TW" to the unsupported "zh".
    { expected: "zh-CN", languages: ["zh-CN"] },
    { expected: "zh-CN", languages: ["zh-CN", "zh", "en"] },
    { expected: "zh-CN", languages: ["zh-CN", "en-US"] },
    { expected: "zh-TW", languages: ["zh-TW", "zh", "en"] },
    { expected: "zh-TW", languages: ["zh-TW", "en"] },
    // Traditional Chinese via script/region subtags (e.g. macOS Safari) must
    // resolve to zh-TW, not be collapsed to Simplified by a generic "zh-*" match.
    { expected: "zh-TW", languages: ["zh-Hant-TW", "en"] },
    { expected: "zh-TW", languages: ["zh-Hant", "en"] },
    { expected: "zh-TW", languages: ["zh-HK", "en"] },
    { expected: "zh-TW", languages: ["zh-MO", "en"] },
    // Simplified markers and a bare "zh" map to zh-CN.
    { expected: "zh-CN", languages: ["zh-Hans-CN", "en"] },
    { expected: "zh-CN", languages: ["zh-SG", "en"] },
    { expected: "zh-CN", languages: ["zh", "en"] },
  ])("keeps Chinese locale $expected for $languages", async ({ expected, languages }) => {
    expect(await resolveLanguage(languages)).toBe(expected);
  });

  it("still honors a genuinely preferred non-English language", async () => {
    expect(await resolveLanguage(["hi"])).toBe("hi");
    expect(await resolveLanguage(["hi-IN", "en"])).toBe("hi");
  });
});
